### ISSUE TITLE
Issue with deleteNode when deleting all nodes

### ISSUE THREAD

**Issue by marcelino-pensa** [OctoMap/octomap#307]
I have implemented a procedure for a fading memory octomap, meaning that it will forget things that it saw very long ago by deleting those nodes using `deleteNode`.
If I end up deleting all nodes from the octree, I noticed that the root node remains alive, and that the octree is saying that my whole volume is occupied.
After I delete all nodes from the octree, only one node remains, and that one is the one whose size is equal to `tree.getResolution() * pow(2.0, 16.0)` (i.e., the root). This node cannot be deleted, as `deleteNodeRecurs` does not delete a node if `node != this->root`. 

My questions are:
- Is there any reason why my octomap reverts back to the root as I delete all nodes? Is this the expected behavior? Why not deleting the root as well? Maybe it can't delete the root because `deleteNodeRecurs` does not delete the root...
- Why can't we delete the root in `deleteNodeRecurs`?

**Comment by marcelino-pensa** [comment]
If my question isn't clear, this is what I mean by deleting all nodes:
```
    for (octomap::OcTree::leaf_iterator it = tree_.begin_leafs(),
                                       end = tree_.end_leafs();
                                       it != end; ++it) {
        key = it.getKey();
        tree_.deleteNode(key, it.getDepth());
    }
```

If we delete all nodes using the method above, there will be one remaining node in the tree that is impossible to delete (the root)

**Comment by ahornung** [comment]
Yes, this seems like an omission that should be fixed. Looking at the [implementation of clear](https://github.com/OctoMap/octomap/blob/devel/octomap/include/octomap/OcTreeBaseImpl.hxx#L512), there's no reason not to delete the root as well.

**Comment by ahornung** [comment]
I haven't tested it yet, but PR #288 should fix this as well.

**Comment by ahornung** [comment]
Looking back at the changelog, this issue was fixed long ago with  #11  (v1.6.0) and it's probably a regression.

**Comment by marcelino-pensa** [comment]
I am running on `v1.9.0`, and the problem still exists

**Comment by xiaofengguorui** [comment]
I have the same issu, and when I delete all nodes,  I try to updateNode failed.

**Comment by ahornung** [comment]
Fixed now, also added a few meaningful unit tests for the behavior.
